File: src/structures.zig
// src/structures.zig

const c = @import("c_api.zig").c;

pub export fn setBlockIfReplaceable(ctx: *c.ServerContext, x: c_short, y: u8, z: c_short, block: u8) void {
    const target = c.getBlockAt(ctx, x, y, z);
    // isReplaceableBlock returns a uint8_t in C (0=false, 1=true)
    if (c.isReplaceableBlock(target) == 0 and target != c.B_oak_leaves) {
        return;
    }
    _ = c.makeBlockChange(ctx, x, y, z, block);
}

pub export fn placeTreeStructure(ctx: *c.ServerContext, x: c_short, y: u8, z: c_short) void {
    // Get a random number for tree height and leaf edges.
    const r = c.fast_rand(ctx);
    const height: u8 = 4 + @as(u8, @truncate(r % 3));

    // Set tree base - replace sapling with log and put dirt below.
    _ = c.makeBlockChange(ctx, x, y - 1, z, c.B_dirt);
    _ = c.makeBlockChange(ctx, x, y, z, c.B_oak_log);

    // Create tree stump.
    var i: u8 = 1;
    while (i < height) : (i += 1) {
        setBlockIfReplaceable(ctx, x, y + i, z, c.B_oak_log);
    }

    // Keep track of leaf corners, determines random number bit shift.
    var t: u8 = 2;

    // First (bottom) leaf layer.
    var leaf_x: i32 = -2;
    while (leaf_x <= 2) : (leaf_x += 1) {
        var leaf_z: i32 = -2;
        while (leaf_z <= 2) : (leaf_z += 1) {
            setBlockIfReplaceable(ctx, @intCast(x + leaf_x), y + height - 3, @intCast(z + leaf_z), c.B_oak_leaves);

            // Randomly skip some corners, emulating vanilla tree shape
            var skip_second_layer = false;
            if ((leaf_x == 2 or leaf_x == -2) and (leaf_z == 2 or leaf_z == -2)) {
                t += 1;
                if ((r >> @truncate(t)) & 1 != 0) {
                    skip_second_layer = true;
                }
            }
            if (!skip_second_layer) {
                setBlockIfReplaceable(ctx, @intCast(x + leaf_x), y + height - 2, @intCast(z + leaf_z), c.B_oak_leaves);
            }
        }
    }

    // Second (top) leaf layer.
    leaf_x = -1;
    while (leaf_x <= 1) : (leaf_x += 1) {
        var leaf_z: i32 = -1;
        while (leaf_z <= 1) : (leaf_z += 1) {
            setBlockIfReplaceable(ctx, @intCast(x + leaf_x), y + height - 1, @intCast(z + leaf_z), c.B_oak_leaves);

            var skip_top_layer = false;
            if ((leaf_x == 1 or leaf_x == -1) and (leaf_z == 1 or leaf_z == -1)) {
                t += 1;
                if ((r >> @truncate(t)) & 1 != 0) {
                    skip_top_layer = true;
                }
            }

            if (!skip_top_layer) {
                setBlockIfReplaceable(ctx, @intCast(x + leaf_x), y + height, @intCast(z + leaf_z), c.B_oak_leaves);
            }
        }
    }
}

File: src/structures.c

#include "globals.h"
#include "tools.h"
#include "registries.h"
#include "worldgen.h"
#include "procedures.h"
#include "structures.h"

void setBlockIfReplaceable (short x, uint8_t y, short z, uint8_t block) {
  uint8_t target = getBlockAt(x, y, z);
  if (!isReplaceableBlock(target) && target != B_oak_leaves) return;
  makeBlockChange(x, y, z, block);
}

// Places a tree centered on the input coordinates
void placeTreeStructure (short x, uint8_t y, short z) {

  // Get a random number for tree height and leaf edges
  uint32_t r = fast_rand();
  uint8_t height = 4 + (r % 3);

  // Set tree base - replace sapling with log and put dirt below
  makeBlockChange(x, y - 1, z, B_dirt);
  makeBlockChange(x, y, z, B_oak_log);
  // Create tree stump
  for (int i = 1; i < height; i ++) {
    setBlockIfReplaceable(x, y + i, z, B_oak_log);
  }
  // Keep track of leaf corners, determines random number bit shift
  uint8_t t = 2;
  // First (bottom) leaf layer
  for (int i = -2; i <= 2; i ++) {
    for (int j = -2; j <= 2; j ++) {
      setBlockIfReplaceable(x + i, y + height - 3, z + j, B_oak_leaves);
      // Randomly skip some corners, emulating vanilla tree shape
      if ((i == 2 || i == -2) && (j == 2 || j == -2)) {
        t ++;
        if ((r >> t) & 1) continue;
      }
      setBlockIfReplaceable(x + i, y + height - 2, z + j, B_oak_leaves);
    }
  }
  // Second (top) leaf layer
  for (int i = -1; i <= 1; i ++) {
    for (int j = -1; j <= 1; j ++) {
      setBlockIfReplaceable(x + i, y + height - 1, z + j, B_oak_leaves);
      if ((i == 1 || i == -1) && (j == 1 || j == -1)) {
        t ++;
        if ((r >> t) & 1) continue;
      }
      setBlockIfReplaceable(x + i, y + height, z + j, B_oak_leaves);
    }
  }

}

// Places an explosion crater centered on the input coordinates
void placeCraterStructure (short x, uint8_t y, short z, uint8_t radius) {
  int radius_sq = radius * radius;
  
  // Air blocks are placed in a sphere
  // Starting with y to check for height bounds
  for (int dy = -radius; dy <= radius; dy++) {
    if (y + dy < 0 || y + dy > 255) continue;
    int dy_sq = dy * dy;
    for (int dx = -radius; dx <= radius; dx++) {
      int dx_sq = dx * dx;
      for (int dz = -radius; dz <= radius; dz++) {
        int dist_sq = dx_sq + dy_sq + (dz * dz);
        if (dist_sq <= radius_sq) {
          // Don't replace what is already air
          uint8_t current_block = getBlockAt(x + dx, y + dy, z + dz);
          if (current_block != B_air) {
            makeBlockChange(x + dx, y + dy, z + dz, B_air);
          }
        }
      }
    }
  }
}
